engines/wintermute/ad/ad_attach_3dx.cpp

```diff
@@ -0,0 +1,271 @@
+ */
+
+#include "engines/wintermute/ad/ad_attach_3dx.h"
+#include "engines/wintermute/base/base_active_rect.h"
+#include "engines/wintermute/base/base_game.h"
+#include "engines/wintermute/base/gfx/opengl/base_render_opengl3d.h"
+#include "engines/wintermute/base/gfx/opengl/shadow_volume.h"
+#include "engines/wintermute/base/gfx/x/modelx.h"
+#include "engines/wintermute/base/scriptables/script.h"
+#include "engines/wintermute/base/scriptables/script_stack.h"
+#include "engines/wintermute/base/scriptables/script_value.h"
+
+namespace Wintermute {
+
+IMPLEMENT_PERSISTENT(AdAttach3DX, false)
+
+//////////////////////////////////////////////////////////////////////////
+AdAttach3DX::AdAttach3DX(BaseGame *inGame, BaseObject *owner) : AdObject3D(inGame), _owner(owner) {
+	_dropToFloor = false;
+}
+
+//////////////////////////////////////////////////////////////////////////
+AdAttach3DX::~AdAttach3DX(void) {
+	_owner = nullptr; // ref only
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool AdAttach3DX::init(const char *modelFile, const char *name, const char *parentBone) {
+	delete _modelX;
+	_modelX = nullptr;
+
+	_parentBone = parentBone;
+	setName(name);
+
+	_modelX = new ModelX(_gameRef, _owner);
+
+	if (!_modelX) {
+		return false;
+	}
+
+	bool res = _modelX->loadFromFile(modelFile);
+	if (!res) {
+		delete _modelX;
+		_modelX = nullptr;
+	}
+
+	return res;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool AdAttach3DX::update() {
+	AdObject3D::update();
+	if (_modelX) {
+		return _modelX->update();
+	} else {
+		return true;
+	}
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool AdAttach3DX::displayAttachable(const Math::Matrix4 &viewMat, bool registerObjects) {
+	BaseRenderOpenGL3D *renderer = (BaseRenderOpenGL3D *)_gameRef->_renderer;
+
+	Math::Matrix4 finalMat = _worldMatrix * viewMat;
+	renderer->pushWorldTransform(finalMat);
+
+	if (_modelX) {
+		_modelX->render();
+	}
+
+	if (registerObjects && _owner && _owner->_registrable) {
+		_gameRef->_renderer->addRectToList(new BaseActiveRect(_gameRef, _owner, _modelX,
+		                                                      _modelX->_boundingRect.left,
+		                                                      _modelX->_boundingRect.top,
+		                                                      _modelX->_boundingRect.right - _modelX->_boundingRect.left,
+		                                                      _modelX->_boundingRect.bottom - _modelX->_boundingRect.top,
+		                                                      true));
+	}
+
+	renderer->popWorldTransform();
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool AdAttach3DX::displayShadowVol(const Math::Matrix4 &modelMat, const Math::Vector3d &light, float extrusionDepth, bool update) {
+	BaseRenderOpenGL3D *renderer = (BaseRenderOpenGL3D *)_gameRef->_renderer;

```

Comment:
Wasn't the plan to do something about these casts?

Reply:
Forgot the c-style casts